src/ImageSharp/Formats/Jpeg/Components/Encoder/HuffmanScanEncoder.cs

```diff
@@ -310,54 +308,59 @@ public void EncodeGrayscale<TPixel>(Image<TPixel> pixels, ref Block8x8F luminanc
-        /// <param name="quantTable">Quantization table provided by the callee.</param>
-        /// <param name="cancellationToken">The token to monitor for cancellation.</param>
-        public void EncodeRgb<TPixel>(Image<TPixel> pixels, ref Block8x8F quantTable, CancellationToken cancellationToken)
+        /// <param name="frame">Frame to encode.</param>
+        /// <param name="converter">Converter from color to spectral.</param>
+        /// <param name="cancellationToken">The token to request cancellation.</param>
+        private void EncodeThreeComponentScanBaselineInterleaved444<TPixel>(JpegFrame frame, SpectralConverter<TPixel> converter, CancellationToken cancellationToken)
             where TPixel : unmanaged, IPixel<TPixel>
         {
-            FastFloatingPointDCT.AdjustToFDCT(ref quantTable);
-
-            this.huffmanTables = HuffmanLut.TheHuffmanLut;
+            int mcusPerColumn = frame.McusPerColumn;
+            int mcusPerLine = frame.McusPerLine;
 
-            // ReSharper disable once InconsistentNaming
-            int prevDCR = 0, prevDCG = 0, prevDCB = 0;
+            Component c2 = frame.Components[2];
+            Component c1 = frame.Components[1];
+            Component c0 = frame.Components[0];
 
-            ImageFrame<TPixel> frame = pixels.Frames.RootFrame;
-            Buffer2D<TPixel> pixelBuffer = frame.PixelBuffer;
-            RowOctet<TPixel> currentRows = default;
+            ref HuffmanLut c0dcHuffmanTable = ref this.dcHuffmanTables[c0.DcTableId];
+            ref HuffmanLut c0acHuffmanTable = ref this.acHuffmanTables[c0.AcTableId];
+            ref HuffmanLut c1dcHuffmanTable = ref this.dcHuffmanTables[c1.DcTableId];
+            ref HuffmanLut c1acHuffmanTable = ref this.acHuffmanTables[c1.AcTableId];
+            ref HuffmanLut c2dcHuffmanTable = ref this.dcHuffmanTables[c2.DcTableId];
+            ref HuffmanLut c2acHuffmanTable = ref this.acHuffmanTables[c2.AcTableId];
 
-            var pixelConverter = new RgbForwardConverter<TPixel>(frame);
+            ref Block8x8 c0BlockRef = ref MemoryMarshal.GetReference(c0.SpectralBlocks.DangerousGetRowSpan(y: 0));
+            ref Block8x8 c1BlockRef = ref MemoryMarshal.GetReference(c1.SpectralBlocks.DangerousGetRowSpan(y: 0));
+            ref Block8x8 c2BlockRef = ref MemoryMarshal.GetReference(c2.SpectralBlocks.DangerousGetRowSpan(y: 0));
 
-            for (int y = 0; y < pixels.Height; y += 8)
+            for (int j = 0; j < mcusPerColumn; j++)
             {
                 cancellationToken.ThrowIfCancellationRequested();
-                currentRows.Update(pixelBuffer, y);
 
-                for (int x = 0; x < pixels.Width; x += 8)
+                // Convert from pixels to spectral via given converter
+                converter.ConvertStrideBaseline();
+
+                // Encode spectral to binary
+                for (int i = 0; i < mcusPerLine; i++)
```

Comment:
```suggestion
                for (nint i = 0; i < (nint)(uint)mcusPerLine; i++)
```
I guess the casts are necessary for the JIT to keep that expression in a register.